File: Python/merge_sort.py
# All O(nLogn)


def merge_sort_naive(fin):
    ''' Mergesort - return sorted (storage problem) 
        each recursion creates a copy
    '''
    if len(fin) < 2:
        return fin

    mid = int(len(fin) / 2)
    left = merge_sort_naive(fin[:mid])
    right = merge_sort_naive(fin[mid:])

    i = j = 0

    fout = []
    while len(fout) < len(fin):
        if j >= len(right) or (i < mid and left[i] < right[j]):
            fout.append(left[i])
            i += 1
        elif j < len(right):
            fout.append(right[j])
            j += 1

    return fout


def merge(fin, fout, start, end):
    ''' Mergesort - space efficient 

    '''
    if end - start < 2:
        return fout
    if end - start == 2:
        if fout[start] > fout[start + 1]:
            fout[start], fout[start + 1] = fout[start + 1], fout[start]
        return fout

    mid = int((start + end) / 2)
    merge(fout, fin, start, mid)
    merge(fout, fin, mid, end)

    i = start
    j = end
    ind = start
    while ind < end:
        if j >= end or (i < mid and fin[i] < fin[j]):
            fout[ind] = fin[i]
            i += 1
        else:
            fout[ind] = fin[j]
            j += 1
        ind += 1

def merge_sort(fin):
    copy = list(fin)
    return merge(copy,  fin, 0, len(fin)-1)

File: Java/MergeSort.java
import java.util.ArrayList;
import java.util.List;

public class MergeSort {
  /**
   * MergeSort v.1
   * it uses naive, split, merge
   */
  public void naive(int[] fin) {
    if (fin.length <= 1)
      return;
    int midInd = fin.length / 2 + fin.length % 2;

    // Creates a fresh temp arrays
    int[] lowerHalf = new int[midInd];
    int[] upperHalf = new int[fin.length - midInd];

    // Splits a given array and sort them recursively
    split(fin, lowerHalf, upperHalf);
    naive(lowerHalf);
    naive(upperHalf);

    // Merge sorted temp arrays to fin
    merge(fin, lowerHalf, upperHalf);
  }

  /** 
  * Divides a given array and copies its content to given temp arrays 
  */
  public void split(int[] fin, int[] lowerHalf, int[] upperHalf) {
    int ind = 0;
    int lowerInd = lowerHalf.length;
    for (int elm : fin) {
      if (ind < lowerInd)
        lowerHalf[ind] = fin[ind];
      else
        upperHalf[ind - lowerInd] = fin[ind];
      ind++;
    }
  }

  /**
  * Merges given temp arrays to input array 
  */
  public void merge(int[] fin, int[] lowerHalf, int[] upperHalf) {
    int mergeInd = 0;
    int lowerHalfInd = 0;
    int upperHalfInd = 0;

    // Copies sorted content from temp arrays to fin
    while (lowerHalfInd < lowerHalf.length && upperHalfInd < upperHalf.length) {
      if (lowerHalf[lowerHalfInd] < upperHalf[upperHalfInd]) {
        fin[mergeInd] = lowerHalf[lowerHalfInd];
        lowerHalfInd++;
      } else if (upperHalfInd < upperHalf.length) {
        fin[mergeInd] = upperHalf[upperHalfInd];
        upperHalfInd++;
      }
      mergeInd++;
    }

    // If some content are left in either of temp arrays, copies those to fin.
    if (lowerHalfInd < lowerHalf.length) {
      while (mergeInd < fin.length) {
        fin[mergeInd++] = lowerHalf[lowerHalfInd++];
      }
    }
    if (upperHalfInd < upperHalf.length) {
      while (mergeInd < fin.length) {
        fin[mergeInd++] = upperHalf[upperHalfInd++];
      }
    }
  }

  /**
   * MergeSort v.1 - Sorting linked list
  */
  public List<Integer> mergeL(List<Integer> lowerHalf, List<Integer> upperHalf, List<Integer> accumulator) {
    if (lowerHalf.isEmpty()) {
      accumulator.addAll(upperHalf);
    } else if (upperHalf.isEmpty()) {
      accumulator.addAll(lowerHalf);
    } else {
      if (lowerHalf.get(0) <= upperHalf.get(0)) {
        accumulator.add(lowerHalf.get(0));
        return mergeL(lowerHalf.subList(1, lowerHalf.size()), upperHalf, accumulator);
      } else {
        accumulator.add(upperHalf.get(0));
        return mergeL(lowerHalf, upperHalf.subList(1, upperHalf.size()), accumulator);
      }
    }
    return accumulator;
  }

  public List<Integer> mergeSort(List<Integer> fin) {
    int mid = fin.size() / 2;
    if (mid == 0)
      return fin;
    return mergeL(mergeSort(fin.subList(0, mid)), mergeSort(fin.subList(mid, fin.size())), new ArrayList<Integer>());
  }
}